xds/src/main/java/io/grpc/xds/XdsClientImpl.java

```diff
@@ -0,0 +1,477 @@
+   * resource name is found. Otherwise, sends back a NACK request. Proceed with the
+   * resolved RouteConfiguration (if exists) to find the VirtualHost configuration for the "xds:"
+   * URI (with the port, if any, stripped off).
+   */
+  private void handleRdsResponse(DiscoveryResponse rdsResponse) {
+    logger.log(Level.FINE, "Received an RDS response: {0}", rdsResponse);
+    List<com.google.protobuf.Any> respResources = rdsResponse.getResourcesList();
+    RouteConfiguration routeConfig = null;
+    try {
+      for (com.google.protobuf.Any res : respResources) {
+        RouteConfiguration conf = res.unpack(RouteConfiguration.class);
+        if (routeConfig.getName().equals(adsStream.rdsResourceName)) {
+          routeConfig = conf;
+          // Ignore remaining route configs once the requested one is found.
+          break;
+        }
+      }
+    } catch (InvalidProtocolBufferException e) {
+      configWatcher.onError(Status.fromThrowable(e).augmentDescription("Invalid RDS response"));
+    }
+    if (routeConfig != null) {
+      // Accept this route configuration update.
+      adsStream.rdsVersion = rdsResponse.getVersionInfo();
+    }
+    // Send an ACK/NACK request.
+    adsStream.sendRdsRequest(adsStream.rdsResourceName, rdsResponse.getNonce());
+    if (routeConfig != null) {
+      processRouteConfig(routeConfig);
+    }
+  }
+
+  /**
+   * Processes RouteConfiguration message, which may either contain a VirtualHost with domains
+   * matching the "xds:" URI directly in-line or contain the configuration to send VHDS requests
+   * for dynamic resolution (currently not implemented). Forwards the resolved VirtualHost
+   * configuration as a {@link io.grpc.xds.XdsClient.ConfigUpdate} to the registered config watcher.
+   * Otherwise, forwards an error message to the config watcher.
+   */
+  private void processRouteConfig(RouteConfiguration config) {
+    List<VirtualHost> virtualHosts = config.getVirtualHostsList();
+    String clusterName = null;
+    // Proceed with the virtual host that has longest wildcard matched domain name with the
+    // original "xds:" URI.
+    int matchingLen = -1;  // longest length of wildcard pattern that matches host name
+    for (VirtualHost vHost : virtualHosts) {
+      for (String domain : vHost.getDomainsList()) {
+        if (matchHostName(hostName, domain) && domain.length() > matchingLen) {
+          matchingLen = domain.length();
+          // The client will look only at the last route in the list (the default route),
+          // whose match field must be empty and whose route field must be set.
+          List<Route> routes = vHost.getRoutesList();
+          if (!routes.isEmpty()) {
+            Route route = routes.get(routes.size() - 1);
+            // TODO(chengyuanzhang): check the match field must be empty.
+            if (route.hasRoute()) {
+              clusterName = route.getRoute().getCluster();
```

Comment:
break? Or support multiple (weighted) clusters.

Reply:
Yes, thanks.

Reply:
Just realized that, this is a longest prefix matching, so we do need to go through all the domains. Anyway, the implementation will change significant as the discussion regarding xDS protocol is going on. Please see latest implementation (later).